### PR TITLE
[C++ API] Improved init API

### LINKED ISSUE #11490: [C++ API] Complete the missing arguments in ray.init(..)
Currently, we support a fraction of the arguments under RayConfig class (`ray/api/ray_config.h`):
```
class RayConfig {
 public:
  WorkerType worker_type = WorkerType::DRIVER;

  RunMode run_mode = RunMode::SINGLE_PROCESS;

  std::string redis_ip;

  int redis_port = 6379;

  std::string redis_password = "5241590000000000";

  int node_manager_port = 62665;

  std::string lib_name = "";

  std::string store_socket = "";

  std::string raylet_socket = "";

  std::string session_dir = "";

  static std::shared_ptr<RayConfig> GetInstance();

 private:
  static std::shared_ptr<RayConfig> config_;
};
``` 

We should support the remaining arguments. 
We should start with min_workers/max_workers at least.
This should also support adding a ray address for example so that we do not start a new head node every time (because we should be able to hook up to an existing one).

### PR DESCRIPTION
**Pull request by SrikarVanavasam**
## Why are these changes needed?

- Added support for more arguments including max_workers and min_workers. 
- Supported adding an address to connect to an existing cluster. 

## Related issue number

Closes #11490

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by AmeerHajAli** [comment]
Thanks for the PR!. @SongGuyang , can you please review this?

**Comment by SongGuyang** [comment]
@AmeerHajAli let me see.

File: cpp/src/ray/api.cc [hunk 1]
```diff
@@ -10,9 +10,11 @@ namespace api {
 std::shared_ptr<RayRuntime> Ray::runtime_ = nullptr;
 
 std::once_flag Ray::is_inited_;
-void Ray::Init() {
-  std::call_once(is_inited_,
-                 [] { runtime_ = AbstractRayRuntime::DoInit(RayConfig::GetInstance()); });
+void Ray::Init(std::string address, bool local_mode, int min_workers, int max_workers) {
```
Comment (SongGuyang) [inline cpp/src/ray/api.cc:13]: I prefer to make `Ray::Init()` having no parameters. You can get any config using `RayConfig::GetInstance()` in anywhere. And in fact in Java API, `Ray.init()` has no parameters too.

File: cpp/src/ray/util/process_helper.cc [hunk 2]
```diff
@@ -45,7 +55,9 @@ static void StopRayNode() {
 void ProcessHelper::RayStart(std::shared_ptr<RayConfig> config,
                              CoreWorkerOptions::TaskExecutionCallback callback) {
   std::string redis_ip = config->redis_ip;
-  if (config->worker_type == WorkerType::DRIVER && redis_ip.empty()) {
+  if (!redis_ip.empty()) {
+    ConnectRayNode(redis_ip, config->redis_password);
```
Comment (SongGuyang) [inline cpp/src/ray/util/process_helper.cc:59]: We already support this in StartRayNode. And in fact, we already support connect a existing cluster which has no test in bazel.

**Comment by AmeerHajAli** [comment]
@SongGuyang @alindkhare , why do you think the consensus here should be?
Should we make this change or use the RayConfig?

**Comment by alindkhare** [comment]
> I prefer to make `Ray::Init()` having no parameters. You can get any config using `RayConfig::GetInstance()` in anywhere. And in fact in Java API, `Ray.init()` has no parameters too.
In general, do we want APIs to be similar to Python or Java?
PS: C++ can have bindings with python

**Comment by SongGuyang** [comment]
@AmeerHajAli @alindkhare In python, you can use ray.init like this:
`ray.init(address="127.0.0.1:6379", include_dashboard=True, num_cpus=6)`
But in C++, you can not do like this. Maybe the API should be:
`Ray::init("127.0.0.1:6379", true, 6, 0, 0, "", "")`
You can not know the clear name of every params. When we add more and more params to the `Ray::init`, the code will be more confusing and hard to maintain.
Maybe we can support only one param which is `RayConfig` class, but we need more discuss for this.

**Comment by kfstorm** [comment]
@alindkhare I think we should align the C++ API with Python in general, but still follow C++ best practice. We believe this is correct for Java API as well.

@SongGuyang I agree with `Ray::init()` with a signal parameter, but let's call it `RayOptions` instead of `RayConfig` and only expose a subset of `RayConfig`. e.g. `worker_type` is for internal use and we shouldn't allow users to change it.

**Comment by kfstorm** [comment]
In Java, `Ray.init()` has no parameters because there are other ways to configure options. In case some of you are not familiar with this part, here is the doc: https://docs.ray.io/en/latest/configure.html#driver-options

**Comment by SongGuyang** [comment]
@kfstorm ok, `RayOptions ` is good for me.

**Comment by AmeerHajAli** [comment]
@SongGuyang @SrikarVanavasam , what is the status of this?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**